noir/aztec_macros/src/lib.rs

```diff
@@ -1601,3 +1601,143 @@ fn event_signature(event: &StructType) -> String {
+            }
+        }
+    }
+
+    // This crate's traits and impls have not yet been resolved, so we look for impls in unresolved_trait_impls.
+    note_types.extend(unresolved_traits_impls.iter()
+        .filter(|trait_impl| trait_impl.trait_path.segments.last().unwrap().0.contents == "NoteInterface")
+        .filter_map(|trait_impl| match &trait_impl.object_type.typ {
+            UnresolvedTypeData::Named(path, _, _) => Some(path.segments.last().unwrap().0.contents.clone()),
+            _ => None,
+        }));
+
+    // We can now generate a version of compute_note_hash_and_nullifier tailored for the contract in this crate.
+    let func = generate_compute_note_hash_and_nullifier(&note_types);
+
+    // And inject the newly created function into the contract.
+
+    // We don't have a reasonable location for the source code of this autogenerated function, so we simply pass an
+    // empty span. This function should not produce errors anyway so this should not matter.
+    let location = Location::new(Span::empty(0), file_id);
+
+    // These are the same things the ModCollector does when collecting functions: we push the function to the
+    // NodeInterner, declare it in the module (which checks for duplicate definitions), and finally add it to the list
+    // on collected but unresolved functions.
+
+    let func_id = context.def_interner.push_empty_fn();
+    context.def_interner.push_function(func_id, &func.def, ModuleId { krate: *crate_id, local_id: module_id }, location);
+
+    let Ok(_) = context.def_map_mut(crate_id).unwrap()
+        .modules_mut()[module_id.0]
+        .declare_function(
+            func.name_ident().clone(), func_id
+        )
+    else {
+        panic!("Failed to declare the autogenerated compute_note_hash_and_nullifier function, likely due to a duplicate definition. This should not happen");
+    };
+
+    collected_functions.iter_mut().find(|fns| fns.file_id == file_id).unwrap().push_fn(module_id, func_id, func.clone());
+
+    Ok(())
+}
+
+fn generate_compute_note_hash_and_nullifier(note_types: &Vec<String>) -> NoirFunction {
+    let function_source = generate_compute_note_hash_and_nullifier_source(note_types);
+
+    let (function_ast, errors) = parse_program(&function_source);
+    if !errors.is_empty() {
+        dbg!(errors.clone());
+    }
+    assert_eq!(errors.len(), 0, "Failed to parse Noir macro code. This is either a bug in the compiler or the Noir macro code");
+
+    let mut function_ast = function_ast.into_sorted();
+    function_ast.functions.remove(0)
+}
+
+fn generate_compute_note_hash_and_nullifier_source(note_types: &Vec<String>) -> String {
+    if note_types.len() == 0 {
+        "
+        unconstrained fn compute_note_hash_and_nullifier(
+            contract_address: AztecAddress,
+            nonce: Field,
+            storage_slot: Field,
+            note_type_id: Field,
+            serialized_note: [Field; 20]
```

Comment:
If you could add a note that the 20 comes from the `MAX_NOTE_FIELDS_LENGTH` value would be great.

Reply:
We could also import `MAX_NOTE_FIELDS_LENGTH` as part of the prelude and just use it here